commandry-core/src/main/java/de/eldoria/commandry/ArgumentParser.java

```diff
@@ -0,0 +1,66 @@
+package de.eldoria.commandry;
+
+import de.eldoria.commandry.annotation.Optional;
+import de.eldoria.commandry.parser.Parser;
+import de.eldoria.commandry.parser.ParserManager;
+import de.eldoria.commandry.parser.SharedParsers;
+import de.eldoria.commandry.util.reflection.ReflectionUtils;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * An ArgumentParser provides parsing of arguments with both shared parsers and
+ * custom parsers.
+ *
+ * @see SharedParsers
+ * @see ParserManager
+ */
+public class ArgumentParser {
+    private final ParserManager sharedParsers = SharedParsers.getManager();
+    private final ParserManager customParserManager = new ParserManager();
+
+    /**
+     * Parses all optional parameter values from a method and returns them in a map. That way,
+     * they don't need to be parsed multiple times. If an argument cannot be parsed, an error may
+     * be thrown. The returned map is immutable and therefore cannot be modified.
+     *
+     * @param method the method to parse the optional parameters for.
+     * @return a map view of the parsed optional arguments.
+     */
+    public Map<String, Object> parseOptionals(Method method) {
+        var map = new HashMap<String, Object>();
+        for (Parameter parameter : method.getParameters()) {
+            var annotation = ReflectionUtils.getAnnotation(Optional.class, parameter);
+            var name = parameter.getName();
+            var type = parameter.getType();
+            annotation.ifPresent(optional -> map.put(name, parse(optional.value(), type)));
+        }
+        return Collections.unmodifiableMap(map);
+    }
+
+    /**
+     * Parses an input string to an object of the given type.
+     *
+     * @param input the input string to parse.
+     * @param type  the type class the string should be parsed to.
+     * @param <T>   the type.
+     * @return an object of the requested type.
+     */
+    public <T> T parse(String input, Class<T> type) {
+        if (sharedParsers.hasParserFor(type)) {
+            return sharedParsers.parse(input, type);
+        } else if (customParserManager.hasParserFor(type)) {
+            return customParserManager.parse(input, type);
+        } else {
+            throw new IllegalStateException("No parser for optional parameter found."); // TODO
```

Comment:
Fehlt nen throw comment in der methode

Reply:
Besser wäre es vermutlich, wenn eine Fehlermeldung dieser Art direkt beim Registrieren eines Befehls geworfen wird. Das würde ich aber alles in einem späteren Schritt machen, in welchem ich mich auf Fehlerbehandlung generell konzentriere.